Daemonset controller staleness detection

michaelasp opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Adds the ability for the daemonset controller to wait to read its own writes.
#### Which issue(s) this PR is related to:
KEP: https://github.com/kubernetes/enhancements/issues/5647
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)


If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-controller-manager: The daemonset controller now defers syncing a DaemonSet object when the controller has not yet observed daemonset or pod writes from the last time the object was synced. This prevents spurious creation of duplicate pods for nodes when the controller's cache is stale. When a sync is deferred for this reason, a `daemonset_controller_stale_sync_skips_total` metric is incremented and a message is logged by the daemonset controller. This behavior can be temporarily disabled by setting the `StaleControllerConsistencyDaemonSet` feature gate to false.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:

- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/5647
```

xigang commented:
In addition to the DaemonSet controller, we also need to enable this feature for several other core controllers, including Deployment, ReplicaSet, and StatefulSet.

related issue: https://github.com/kubernetes/kubernetes/issues/129795 😄

xigang commented:
1. It’s recommended to split the client-go changes and the DaemonSet Controller logic into two separate PRs.
2. Would using `ConsistencyStore` provide better extensibility? Could it be reused not only for DaemonSet but also for Deployment, ReplicaSet, and StatefulSet?

liggitt commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -171,13 +192,25 @@ func NewDaemonSetsController(
 	daemonSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
+			ds := obj.(*apps.DaemonSet)
+			dsc.consistencyStore.ReadAt(
```
I'm not seeing a spot where we record WroteAt for DaemonSet when updating status (added that in https://github.com/liggitt/kubernetes/tree/pocConsistentDs)

michaelasp commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -336,8 +385,29 @@ func (dsc *DaemonSetsController) processNextWorkItem(ctx context.Context) bool {
+	klog.InfoS("Checking if ready")
+	if isReady, err := dsc.consistencyStore.IsReady(dsNamespacedName); !isReady && err == nil {
+		klog.InfoS("Is not ready")
+		dsc.queue.AddRateLimited(dsKey)
```
That's fair, I think we can just return in that case.

liggitt commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -203,16 +236,25 @@ func NewDaemonSetsController(
+			dsc.consistencyStore.ReadAt(podGVK, pod.ResourceVersion)
 			dsc.updatePod(logger, oldObj, newObj)
 		},
 		DeleteFunc: func(obj interface{}) {
```
they *should* be returning the object from the cache with the resourceVersion of the delete

serathius replied:
Sorry for confusion, I was wrong about RV and @liggitt is right. Apiserver overrides the RV for the delete objects. Delete event on watch will come with object that was deleted, but with a RV of the deletion. So the RV from deletes is also useful for our purposes.

michaelasp commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -336,8 +385,29 @@ func (dsc *DaemonSetsController) processNextWorkItem(ctx context.Context) bool {
+	// but we don't risk reconciling forever. This will only happen if some
+	// resource version parsing is not working properly.
+	klog.InfoS("Checking if ready")
+	if isReady, err := dsc.consistencyStore.IsReady(dsNamespacedName); !isReady && err == nil {
```
Added this, it shouldn't really matter where we do this portion.

michaelasp commented:
Added some tests and the feature gate, still needs some work but let's get the tests running :)

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -132,9 +132,11 @@ func (f *RealFIFO) hasSynced_locked() bool {
-	_, err := f.keyOf(obj)
-	if err != nil {
-		return KeyError{obj, err}
+	if deltaActionType != Bookmark {
```
Yep, moved!

michaelasp commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -203,16 +236,25 @@ func NewDaemonSetsController(
 
 	// Watch for creation/deletion of pods. The reason we watch is that we don't want a daemon set to create/delete
 	// more pods until all the effects (expectations) of a daemon set's create/delete have been observed.
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
```
Done

michaelasp commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -171,13 +192,25 @@ func NewDaemonSetsController(
+		consistencyStore: consistencyStore,
 	}
 
+	dsGVK := schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "DaemonSet"}
```
Done

michaelasp commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,185 @@
+	"k8s.io/klog/v2"
+)
+
+type ConsistencyStore struct {
```
Add tests to the consistency store.

liggitt commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -288,6 +330,13 @@ func (dsc *DaemonSetsController) deleteDaemonset(logger klog.Logger, obj interfa
 		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", ds, err))
 		return
 	}
+	dsc.consistencyStore.Clear(
```
I think this order could have happened:

1. handler thread observes create of DS foo, adds to sync queue
2. sync loop thread starts running syncDaemonSet on DS foo from queue in step 1
3. handler thread observes delete of DS foo, calls Clear, deletes from consistencyStore, adds to sync queue
4. sync loop thread writes a pod for DS foo, calls WroteAt and recreates record for foo in consistencyStore, attempts to write DS status and fails because it's not found, completes syncDaemonSet
5. sync loop thread starts running syncDaemonSet on DS foo from queue in step 3, gets NotFound

if we didn't also call Clear from the sync loop in step 5, we'll be left with a dangling record, since the delete event was already observed and triggered the Clear

liggitt commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,188 @@
+package util
```
add copyright boilerplate headers to new files

liggitt commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -1202,10 +1282,28 @@ func (dsc *DaemonSetsController) syncDaemonSet(ctx context.Context, key string)
+		dsc.consistencyStore.Clear(
+			dsNamespacedName,
+			ds.UID,
+		)
```
this sync loop is the only thing that was going to write to this record, and it couldn't find the daemonset, we should unconditionally delete any entry regardless of uid:

```suggestion
		dsc.consistencyStore.Clear(dsNamespacedName, "")
```

michaelasp committed:
Daemonset Consistency

Add the ability for the daemonset controller to figure out whether it has read its own writes for pods and daemonset objects.

michaelasp commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -1202,10 +1282,28 @@ func (dsc *DaemonSetsController) syncDaemonSet(ctx context.Context, key string)
+	// If our writes have not yet been observed by our informers, we should not
+	// continue since our reads will not be in sync with our previously enacted
+	// state.
+	if utilfeature.DefaultFeatureGate.Enabled(features.StaleControllerConsistency) && !dsc.consistencyStore.IsReady(dsNamespacedName) {
```
Yeah was thinking this as well, in that case we only need to have the feature gate call where we instantiate the consistency store. Not sure whether to throw it into a different file.

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -238,6 +240,7 @@ type ResourceEventHandler interface {
 	OnAdd(obj interface{}, isInInitialList bool)
 	OnUpdate(oldObj, newObj interface{})
 	OnDelete(obj interface{})
+	OnBookmark(resourceVersion string)
```
Yeah, this is a useful method for many consumers of the interface and is necessary for implementing callbacks for bookmark events in the cache. I don't see much of a way around adding the method and since it's a relatively simple method I think it's valid to add.

michaelasp commented:
/cc @janetkuo
For sig-apps approval for daemonset controller changes. Should be a noop unless feature gate is enabled.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +260,12 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+func (f *RealFIFO) Bookmark(resourceVersion string) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	return f.addToItems_locked(Bookmark, false, resourceVersion)
```
we *do* want to skip transform for bookmark events

```suggestion
	return f.addToItems_locked(Bookmark, true, resourceVersion)
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -52,6 +52,8 @@ type DeltaFIFOOptions struct {
 	// When false, `Sync` events will be sent instead.
 	EmitDeltaTypeReplaced bool
 
+	EmitDeltaTypeBookmark bool
```
add godoc describing what this does and what the "object" in the emitted deltas is (string resourceVersion)

michaelasp replied:
Done

liggitt commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -180,6 +184,8 @@ const (
 	// Sync is for synthetic events during a periodic resync.
 	Sync DeltaType = "Sync"
+
+	Bookmark DeltaType = "Bookmark"
```
godoc

liggitt commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,240 @@
+	"k8s.io/apimachinery/pkg/util/resourceversion"
+)
+
+type ConsistencyStore interface {
```
add godoc for the methods, clarifying which ones raise the high-water rv (I think all of them), how set and unset ownerUID is used by Clear(), what IsReady does, etc

michaelasp replied:
Added godoc to interface and actual implementation

liggitt commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -1210,10 +1297,25 @@ func (dsc *DaemonSetsController) syncDaemonSet(ctx context.Context, key string)
+	// continue since our reads will not be in sync with our previously enacted
+	// state.
+	if !dsc.consistencyStore.IsReady(dsNamespacedName) {
+		logger.Info("Daemonset is not ready for reconciliation, previous writes have not been observed in the cache", "ds", dsNamespacedName)
```
I think we may want to switch from `IsReady(namespacedName) bool` to `EnsureReady(namespacedName) error` so we can return details about the group/resource/expectedRV/actualRV that blocked readiness and log it here

liggitt reviewed (commented):
getting close, but also getting close to code freeze 🥶 

also needs rebase / reconciliation with changes from https://github.com/kubernetes/kubernetes/pull/132240

liggitt commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -277,6 +289,9 @@ func (f *RealFIFO) Replace(newItems []interface{}, resourceVersion string) error
 	queuedKeys := []string{}
 	lastQueuedItemForKey := map[string]Delta{}
 	for _, queuedItem := range queuedItems {
+		if queuedItem.Type == Bookmark {
```
oh, I misread this ... I thought this was skipping bookmarks in the `newItems []interface{}`

yeah, dropping queued bookmarks out is fine

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -101,6 +101,10 @@ func (f *RealFIFO) keyOf(obj interface{}) (string, error) {
 		obj = d.Newest().Object
```
I think it's better to keep below since we use it for tests as well and directly pass the delta into it, otherwise we have to rewrite some tests. Lmk what you think though, I don't mind too much either way but don't see much point to refactoring the tests for this.

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -358,6 +373,11 @@ func (f *RealFIFO) Replace(newItems []interface{}, resourceVersion string) error
 		f.initialPopulationCount = len(f.items)
 	}
 
+	retErr := f.addToItems_locked(Bookmark, false, resourceVersion)
```
Done

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -145,7 +151,7 @@ func (f *RealFIFO) addToItems_locked(deltaActionType DeltaType, skipTransform bo
 	if f.transformer != nil {
 		_, isTombstone := obj.(DeletedFinalStateUnknown)
-		if !isTombstone && !skipTransform {
+		if !isTombstone && !skipTransform && deltaActionType != Bookmark {
```
Done

michaelasp commented:
/cc @sanposhiho 

For trivial approval on scheduler changes(just added a new empty method for tests)

sanposhiho reviewed (commented):
/approve

only sig-scheduling part

liggitt commented on pkg/controller/controller_utils.go:
sweeping through other writes RealPodControl does, it looks like we don't have RV info for Delete

that led to looking at what callers of Delete are doing, and it looks like some might be clearing expectations on *any* error ... in a separate PR, we should consider sweeping and narrowing that to just NotFound errors:

```
diff --git a/pkg/controller/daemon/daemon_controller.go b/pkg/controller/daemon/daemon_controller.go
index 8fbe4391b01..6921b7e240e 100644
--- a/pkg/controller/daemon/daemon_controller.go
+++ b/pkg/controller/daemon/daemon_controller.go
@@ -1116,8 +1116,9 @@ func (dsc *DaemonSetsController) syncNodes(ctx context.Context, ds *apps.DaemonS
 		go func(ix int) {
 			defer deleteWait.Done()
 			if err := dsc.podControl.DeletePod(ctx, ds.Namespace, podsToDelete[ix], ds); err != nil {
-				dsc.expectations.DeletionObserved(logger, dsKey)
-				if !apierrors.IsNotFound(err) {
+				if apierrors.IsNotFound(err) {
+					dsc.expectations.DeletionObserved(logger, dsKey)
+				} else {
 					logger.V(2).Info("Failed deletion, decremented expectations for daemon set", "daemonset", klog.KObj(ds))
 					errCh <- err
 					utilruntime.HandleError(err)
```

michaelasp replied:
+1, I can look at adding that in a follow up.

liggitt commented on pkg/features/kube_features.go:
```diff
@@ -1766,6 +1773,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
+	StaleControllerConsistency: {
+		{Version: version.MustParse("1.35"), Default: false, PreRelease: featuregate.Alpha},
```
let's make this specific to StaleControllerConsistencyDaemonSet, so we can mitigate per-type if needed:

```suggestion
		{Version: version.MustParse("1.36"), Default: false, PreRelease: featuregate.Alpha},
```

liggitt replied:
we should *consider* making this beta / enabled-by-default, since it's effectively a bugfix for a runaway controller failure mode

michaelasp replied:
I'm not opposed, we still have time in the release to get a good amount of runs with the changes. If we see any issues it should likely pop up soon, and the feature gates it relies on are all in beta. I'll update it for now but would like some discussion to see where it should land.

liggitt commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,266 @@
+	// writes is a map of owner -> ownerRecord
+	writes map[types.NamespacedName]*ownerRecord
+
+	stores map[schema.GroupResource]cache.Store
```
let's make the interface of the value just `LastStoreSyncResourceVersion() string` (can make a local interface if you want, or put an interface in the cache package)

liggitt commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,266 @@
+	return nil
+}
+
+type resourceVersions struct {
```
we also don't expect *any* cross-worker contention on an ownerRecord WroteAt / EnsureReady calls, so I think we can dramatically simplify the fields / accesses to take a lock, compare, and write to the map if higher:

```go
type ownerRecord struct {
	// ownerUID must not be mutated after creation
	ownerUID types.UID

	versionsLock sync.Mutex
	versions     map[schema.GroupResource]string
```

liggitt commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,266 @@
+			return &ConsistencyError{
+				WroteRV:   owner.String(),
+				GroupKind: gk.String(),
+			}
```
```suggestion
			// read version is not as new as wrote version, not ready
			return &ConsistencyError{
				ReadRV: readRV,
				WroteRV:   wroteRV,
				GroupResource: gr.String(),
			}
```

liggitt commented on pkg/controller/daemon/metrics/metrics.go:
```diff
@@ -0,0 +1,47 @@
+			StabilityLevel: metrics.ALPHA,
+		},
+		// These are the labels (dimensions)
+		[]string{"namespace", "daemonset", "stale_object_group_kind"},
```
I don't think we want namespace/name as labels

Split `group`,`resource` into distinct labels

I'm not sure if we want a metric per controller, or to make the controller name a label. @richabanker might have input. It seems like most things have per-controller metrics.

liggitt commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -1212,10 +1272,33 @@ func (dsc *DaemonSetsController) syncDaemonSet(ctx context.Context, key string)
+	// continue since our reads will not be in sync with our previously enacted
+	// state.
+	if err := dsc.consistencyStore.EnsureReady(dsNamespacedName); err != nil {
+		logger.Error(err, "Daemonset is not ready for reconciliation", "ds", dsNamespacedName)
```
drop this logging, we'll return the consistency error which will get logged at a higher level

michaelasp replied:
Done

michaelasp commented on pkg/features/kube_features.go:
```diff
@@ -2374,6 +2385,8 @@ var defaultKubernetesFeatureGateDependencies = map[featuregate.Feature][]feature
 
 	SidecarContainers: {},
 
+	StaleControllerConsistency: {},
```
Done, I had to re order the feature gate initialization however since the dependency checker wouldn't pick up client-go feature gates otherwise.

michaelasp commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2025 The Kubernetes Authors.
```
Done!

michaelasp commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,266 @@
+func (w *ownerRecord) EnsureReady(c *RealConsistencyStore) error {
+	w.versions.versionsLock.RLock()
+	defer w.versions.versionsLock.RUnlock()
+	for gk, owner := range w.versions.versions {
```
Updated, thanks

michaelasp commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -1212,10 +1272,33 @@ func (dsc *DaemonSetsController) syncDaemonSet(ctx context.Context, key string)
+			metrics.DaemonsetRequeueSkips.WithLabelValues(
+				namespace,
+				name,
+				consistencyErr.GroupKind,
```
Updated, also now splits group and resource to be more in line with other metrics at @richabanker's suggestion. I think version is hard to get here and doesn't make sense.

michaelasp commented on pkg/controller/daemon/metrics/metrics.go:
```diff
@@ -0,0 +1,48 @@
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const DaemonControllerSubsystem = "daemon_controller"
```
Yep, looks like others use underscores but changed to `daemonset_controller`

michaelasp commented on pkg/controller/daemon/util/consistency.go:
```diff
@@ -0,0 +1,220 @@
+func (c *ConsistencyError) Error() string {
+	if c.ReadRV == "" {
+		return fmt.Sprintf("unable to obtain read value for group resource: %s", c.GroupResource.String())
+	}
```
Yep, that's correct. Removed.

michaelasp commented on pkg/features/kube_features.go:
```diff
@@ -935,6 +935,13 @@ const (
+	// owner: @michaelasp
+	// kep: http://kep.k8s.io/5647
+	//
+	// Introduces the ability for certain controllers to be able to read their writes
```
Updated docs
